NAPS2.Lib/Dependencies/DownloadController.cs

```diff
@@ -97,80 +53,134 @@ public void Stop()
 
     public event EventHandler? DownloadProgress;
 
-    private void StartNextDownload()
+    private async Task<MemoryStream?> TryDownloadQueueItemAsync(string filename, string url)
     {
-        if (_hasError)
+        CurrentFileProgress = 0;
+        CurrentFileSize = 0;
+        DownloadProgress?.Invoke(this, EventArgs.Empty);
+        try
         {
-            var prev = _filesToDownload[FilesDownloaded];
-            Directory.Delete(prev.TempFolder!, true);
-            if (_cancel)
+            var response = await _client.GetAsync(url, HttpCompletionOption.ResponseHeadersRead);
+            response.EnsureSuccessStatusCode();
+            CurrentFileSize = response.Content.Headers.ContentLength.GetValueOrDefault();
+
+            using var contentStream = await response.Content.ReadAsStreamAsync();
+
+            var result = new MemoryStream();
+            long previousLength;
+            byte[] buffer = new byte[1024 * 40];
+            do
             {
-                return;
+                previousLength = result.Length;
+                int length = await contentStream.ReadAsync(buffer, 0, buffer.Length);
+                if (length > 0)
+                {
+                    result.Write(buffer, 0, length);
+                    CurrentFileProgress = result.Length;
+                    DownloadProgress?.Invoke(this, EventArgs.Empty);
+                }
+                if (_cancel)
+                {
+                    throw new OperationCanceledException();
+                }
             }
-            // Retry if possible
-            _urlIndex++;
-            _hasError = false;
+            while (previousLength < result.Length);
+            return result;
+        }
+        catch (OperationCanceledException)
+        {
+            throw;
         }
-        else
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "Error downloading file: {FileName}", filename);
+            return null;
+        }
+    }
+
+    private async Task<MemoryStream?> TryDownloadQueueItemAsync(QueueItem fileToDownload)
+    {
+        foreach (var url in fileToDownload.DownloadInfo.Urls)
         {
-            _urlIndex = 0;
+            MemoryStream? result;
+            try
+            {
+                result = await TryDownloadQueueItemAsync(fileToDownload.DownloadInfo.FileName, url);
+            }
+            catch (OperationCanceledException)
+            {
+                throw;
+            }
```

Comment:
```suggestion
            result = await TryDownloadQueueItemAsync(fileToDownload.DownloadInfo.FileName, url);
```

Okay this is silly... there is no need to catch and immediately rethrow this exception. I will remove it.